android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -39,8 +41,52 @@ class TranslationsMiddleware(
     ) {
         // Pre process actions
         when (action) {
+            // Initially adding the tab
+            is TabListAction.AddTabAction -> {
```

Comment:
Between `AddTabAction` (new tab), `RestoreAction` (reopening after a shutdown), and `SelectTabAction` (reopening after a shutdown and changing tabs), all the possibilities seem to be covered.

This bug is just checking for if the engine is supported, since it'll be a very early requirement. (We won't show entryways into the feature if the engine isn't supported.) 

We had a conversation about this style flow over [here](https://github.com/mozilla-mobile/firefox-android/pull/5224).

[1873672](https://bugzilla.mozilla.org/show_bug.cgi?id=1873672) will probably use these same actions to initialize the rest of the translations data needs.  I'd like to get a stronger sense of how well it works for this case alone first.

Reply:
Hey, @MatthewTighe and @gabrielluong, what do you think about this? Is there anything else I should consider or do you think it is good to start testing this option?

Reply:
Based on the amount of research you did, I feel like this is good enough. I don't really have enough context on these actions to provide a better suggestion. 

My understanding of what we are trying to achieve here is to make sure all the tabs have this engine supported property set. The only alternative I could see is calling this on-demand for the individual tab that is displayed instead of doing this for all the tabs. Do you have a sense of whether or not this operation is expensive? For instance, what happens if an user tries to add/restore 50, 100 or 1000 tabs?

Reply:
Do we need to check if the engine is supported per tab? 

I was re-reading the description for one of the properties below:
> @property isEngineSupported If the engine supports translations on this device.
and was wondering if we could just set this only once since this seem to indicate whether or not the engine supports translations on the device, which seems like a global state and not based on per tab.

Reply:
+1 to this question about whether the state is global. though it is useful on the one hand to have all the translation state encapsulated in `TranslationState`, any global state could definitely move to the top-level browser state or even a `BrowserState.translationState`. If you went the latter route, the current `TranslationState` may make more sense as `SessionTranslationState`

Reply:
Thanks for taking a look!

> Do we need to check if the engine is supported per tab?

Not necessarily, this won't change once found.

>Does the translation engine settings need to fetched per tab or do we only need to know if the engine is ready?

This `isEngineSupported` could be a global setting. 100% it is based on the users device and my understanding is it won't change, so once we know, we know.  (Only exception might be something like a possible firmware update, but I think that would be an extreme example an generally not easily changed - [more details](https://mozilla.slack.com/archives/C04DGR18D0F/p1701376615035419).)

The engine automatically becoming ready is fully managed by Gecko AFAIK currently. They are also managing the new single process [1815339](https://bugzilla.mozilla.org/show_bug.cgi?id=1815339).

LMK how we normally handle something like this and I can refactor for sure!

Reply:
>If you went the latter route, the current TranslationState may make more sense as SessionTranslationState

That could be interesting because supported languages are also not directly tied to the current tab. I'd say download state would probably make sense there too. 

(I think there could also be an argument made for the never translate site and languages global settings to be on the global state, but they have a lot of coupling with the page settings, so they probably shouldn't be on a global state and stay tied to one for tabs/sessions.)

Do you have an example of something similar to this I could pattern off of?

Reply:
Just looking through `BrowserState`, there is global extension state and then tab-specific extension state in `TabSessionState`. I'm not really familiar with extensions, but I would be happy to chat through any implementation details if you'd like

Reply:
Awesome, thanks! I'll try following it and see where I end up. (Interestingly, I had to follow a lot of extensions patterns on the GeckoView side as well, so that makes sense too!)